### PR TITLE
Upstream RDK adapter

### PR DESCRIPTION
**Pull request by kkanag314**

### CODE REVIEW

**Comment by mhughesacn** [comment]
Hi Krishna: Who created the new file open_cdm_adapter.cpp (i.e. RDK/Metrological/Comcast etc., not the contracting company) - please confirm whose copyright applies.
Thank you.

**Comment by DouglasAdler** [comment]
@mhughesacn The rdk adapter already has a copyright header.

* Copyright 2020 RDK Management

This file was created by Comcast but is already used by the open source RDK licensees.

**Comment by mhughesacn** [comment]
Hi Doug: In this component, the default header is Metrological Apache, but Comcast Apache would be OK too as it is your file.  However, there are some RDK Apache headers already, so I don't think it fair to block.

**Comment by DouglasAdler** [comment]
Thanks,
I had this header in there on the guidance of RDK-M.  I'm happy to change it to what the current component uses as it's now part of this repo.
    On Tuesday, March 26, 2024 at 09:24:44 AM CDT, mhughesacn ***@***.***> wrote:  
 
 


Hi Doug: In this component, the default header is Metrological Apache, but Comcast Apache would be OK too as it is your file. However, there are some RDK Apache headers already, so I don't think it fair to block.

File: Source/ocdm/CMakeLists.txt [hunk 1]
```diff
@@ -129,6 +129,14 @@ if (${CDMI_ADAPTER_IMPLEMENTATION} STREQUAL "broadcom-svp-secbuf")
             )
 endif()
 
+if (${CDMI_ADAPTER_IMPLEMENTATION} STREQUAL "rdk")
+    add_definitions( -DUSE_SVP=1 )
```
Comment (sramani-metro) [inline Source/ocdm/CMakeLists.txt:133]: This definition seems redundant as it is not used anywhere

File: Source/ocdm/adapter/open_cdm_adapter.h [hunk 2]
```diff
@@ -113,6 +113,15 @@ EXTERNAL uint32_t opencdm_destruct_session_private(struct OpenCDMSession* sessio
...
 
+/**
+ * \brief adds SVP related features to the caps structure (only if needed by the platform)
+ *
+ * \param caps GstCaps to be updated
+ *
+ * \return Zero on success, non-zero on error.
+ */
+
+    EXTERNAL OpenCDMError opencdm_gstreamer_transform_caps(GstCaps** caps);
```
Comment (sramani-metro) [inline Source/ocdm/adapter/open_cdm_adapter.h:124]: Why is this needed in the adapter? This method seems to be specific to RDK adapter only.

File: Source/ocdm/adapter/rdk/open_cdm_adapter.cpp [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+//#define LOGT(X) LogTelemetry(__FILE__, __LINE__, X)
+//#define LOGE(...) Log(__FILE__, __LINE__, CDMi::ERROR,  __VA_ARGS__)
+
+//#define LOGW(...) Log(__FILE__, __LINE__, CDMi::WARNING,  __VA_ARGS__)
+//#define LOGI(...) Log(__FILE__, __LINE__, CDMi::INFO, __VA_ARGS__)
+//#define LOGD(...) Log(__FILE__, __LINE__, CDMi::DEBUG,  __VA_ARGS__)
+//#define LOGV(...) Log(__FILE__, __LINE__, CDMi::VERBOSE,  __VA_ARGS__)
+
+typedef uint32_t SEC_SIZE;
+struct My_Sec_OpaqueBufferHandle_struct
```
Comment (sramani-metro) [inline Source/ocdm/adapter/rdk/open_cdm_adapter.cpp:42]: Nit: Could choose a better name without My_*

File: Source/ocdm/adapter/rdk/open_cdm_adapter.cpp [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+    return true;
+}
+
+uint32_t opencdm_construct_session_private(struct OpenCDMSession* session, void* &pvtData)
+{
+    bool success = gst_svp_ext_get_context(&pvtData, Server, (unsigned int)session);
+    char buf[25] = { 0 };
+    snprintf(buf, 25, "%X", (unsigned int)session);
+    ASSERT(session != nullptr);
+    session->SetParameter("rpcId", buf);
```
Comment (sramani-metro) [inline Source/ocdm/adapter/rdk/open_cdm_adapter.cpp:81]: Is this SetParameter still needed. I remember from our discussions that the session id string can be used as a unique identifier between the client side and the server side for these purposes.

**Review by sramani-metro — COMMENTED** [review]
Would like to get some things clarified as mentioned in the comments.

File: Source/ocdm/adapter/rdk/open_cdm_adapter.cpp (hunk 4 above)
Comment (kkanag314) [reply Source/ocdm/adapter/rdk/open_cdm_adapter.cpp:81]: This is required for realtek platforms

Comment (sramani-metro) [reply Source/ocdm/adapter/rdk/open_cdm_adapter.cpp:59]: Would like to get this clarified.

Comment (kkanag314) [reply Source/ocdm/adapter/rdk/open_cdm_adapter.cpp]: @sramani-metro @pwielders 
I have modified the rdk-adapter with following changes to take session if as a unique identifier for socket path creation. https://github.com/rdkcentral/ThunderClientLibraries/pull/250/commits/ce8d1981491ecbe57451801da39a55e56035ce2b 
But In client side, If we try to create a context in mediakeysession constructor, the client is trying to establish connection before the server side is ready which is causing the race condition. 
` CWVMediaKeySession::CWVMediaKeySession 
{
......
#ifdef USE_SVP
  if (!m_sessionId.empty()) {
    LOGW("Initializing SVP context for client side ID = %s\n", m_sessionId.c_str());
    gst_svp_ext_get_context(&m_pSVPContext, Client, m_sessionId.c_str());
  }
......
}`

Similarly we tried the same in first decrypt call which is causing issue due to race condition. 
` CDMi_RESULT CWVMediaKeySession::Decrypt(
        uint8_t*                 inData,          // Incoming encrypted data
        const uint32_t           inDataLength,    // Incoming encrypted data length
        uint8_t**                outData,         // Outgoing decrypted data
        uint32_t*                outDataLength,   // Outgoing decrypted data length
        const SampleInfo*        sampleInfo,      // Information required to decrypt Sample
        const IStreamProperties* properties)
{
#ifdef USE_SVP
  if (!m_svpContextInit && !m_sessionId.empty()) {
    LOGW("Initializing SVP context for client side ID = %s\n", m_sessionId.c_str());
    if (gst_svp_ext_get_context(&m_pSVPContext, Client, m_sessionId.c_str()))
	{
		m_svpContextInit = true;
	}

  }
  if (!m_pSVPContext)
  {
    LOGW("Failed to initialize SVP context\n");
  }
#endif
........
}`

Due to this we have tried to invoke the context creation in client side once it is done in server side using setParameter() API. Since setParameter() is already part of ThunderInterfaces Can we use the API for the connection establishment between server and client side. Please share your input regarding this

Comment (pwielders) [reply Source/ocdm/adapter/rdk/open_cdm_adapter.cpp]: Hi @kkanag314, @sramani-metro and myself discussed it. 
First an observation we like to share, how can it be that a less privileged process (the client, e.g. Netflix, that includes the OCDM client library, is a server, the client now needs the right to be able to create server sockets. I think a very undesirable situation, I would have left this up to the OCDM module, which is a more privileged process. So architecturally I think there is something terrible wrong here.
Having said that, suppose we need to cope with this situation, there is, we think, still a way forward. First of all the setParameter is not "already" part of the ThunderInterface, it is this pull request that would bring it in and we are in discussion if it is really needed. We prefer to keep the interface as slim as possible and hence why we do not see the actual use of this method (on the interface) yet.
What we (or ate least what @sramani-metro explained me) is that after the session has been created (OCDM server size), the User private data for the session is attached to it (client side), during this attaching of the User private data client side, the server for Realtek, on client side is started. The socket, with the SessionId is now open, the OCDM server side session now needs to hook up to the server, running client side.
Our remark here would be, why not, after you started the realtek server, call Load() on the Session interface, this call is forwarded to the OCDMServer, implemented in the DRM module side and thus, it would be perfect timing to than connect, in this implementation to the socket (SessionId).
I learned that @sramani-metro reported in ill, so get well soon, but if he is up and running again, he can definitely help you out with the details.
The Load call on OCDMclient library side I am referring to is this:
https://github.com/rdkcentral/ThunderClientLibraries/blob/master/Source/ocdm/open_cdm.h#L443 you should be able to call it from within the adapter implementation when you started the server succefully. And it ends up here: https://github.com/rdkcentral/OCDM-Clearkey/blob/master/MediaSession.h#L47 and here in code https://github.com/rdkcentral/OCDM-Clearkey/blob/master/MediaSession.cpp#L138. It intention is indeed to prepare after the creation of the session to "load" anything reuired to make the session viable for handling. Guess it would be a good point to also hookup to the server instance than.
Let us now if this works for you as well...

**Review by sramani-metro — DISMISSED** [review]
Looks good to me.

**Comment by kkanag314** [comment]
@pwielders Could you please review the changes

**Review by DouglasAdler — APPROVED** [review]

**Review by pwielders — APPROVED** [review]

**Review by sramani-metro — APPROVED** [review]
Looks good.
